python/ray/tune/impl/tuner_internal.py

```diff
@@ -123,6 +126,54 @@ def __init__(
         trainable_ckpt = os.path.join(self._experiment_checkpoint_dir, _TRAINABLE_PKL)
         with open(trainable_ckpt, "wb") as fp:
             pickle.dump(self._trainable, fp)
+        self._maybe_warn_resource_contention()
+
+    def _expected_concurrency(self, cpus_per_trial, cpus_total):
+        num_samples = self._tune_config.num_samples
+        if num_samples < 0:  # simplify this in Tune
+            num_samples = math.inf
+        concurrent_trials = self._tune_config.max_concurrent_trials or 0
+        if concurrent_trials < 1:  # simplify this in Tune
+            concurrent_trials = math.inf
+
+        actual_concurrency = min(
+            (cpus_total // cpus_per_trial, num_samples, concurrent_trials)
+        )
+        return actual_concurrency
+
+    def _maybe_warn_resource_contention(self):
+        scaling_config = None
+        get_scaling_config = getattr(self._trainable, "base_scaling_config")
+        if callable(get_scaling_config):
+            scaling_config = get_scaling_config()
+
+        if not isinstance(scaling_config, ScalingConfig):
+            return
+
+        if getattr(scaling_config, "_max_cpu_fraction_per_node"):
+            return
+
+        cpus_per_trial = scaling_config.total_resources.get("CPU", 0)
+        cpus_total = ray.cluster_resources().get("CPU", 1) + 0.001  # avoid div by 0
```

Comment:
we should use `available_resources` here right?

Reply:
good call, fixed.